From: Filip Pokryvka <fpokryvk@redhat.com>
Date: Mon, 30 Oct 2023 15:50:13 +0200
Subject: [PATCH] netdevsim: set device perm addr

diff --git a/drivers/net/netdevsim/netdev.c b/drivers/net/netdevsim/netdev.c
index 2908e0a0d6e1..16e1621bea8b 100644
--- a/drivers/net/netdevsim/netdev.c
+++ b/drivers/net/netdevsim/netdev.c
@@ -264,6 +312,7 @@ static void nsim_setup(struct net_device *dev)
 {
 	ether_setup(dev);
 	eth_hw_addr_random(dev);
+	memcpy(dev->perm_addr, dev->dev_addr, dev->addr_len);
 
 	dev->tx_queue_len = 0;
 	dev->flags |= IFF_NOARP;
-- 
2.26.2
